ocaml/src/blob_access.ml

```diff
@@ -0,0 +1,384 @@
+    let file = get_file_name fnr in
+    let dir = Bytes.create 20 in
+    let rec fill off1 off2 =
+      if off1 < 20
+      then begin
+        Bytes.set dir off1 file.[off2];
+        if off2 mod 2 = 1
+        then begin
+          if off1 <> 19
+          then Bytes.set dir (off1 + 1) '/';
+          fill (off1 + 2) (off2 + 1)
+        end else fill (off1 + 1) (off2 + 1)
+      end
+    in
+    fill 0 0;
+    let path =
+      String.concat
+        Filename.dir_sep
+        [ files_path; dir; file ] in
+    dir, file, path
+end
+
+type fnr = int64
+type offset = int
+type length = int
+
+class virtual directory_access =
+object
+  method virtual delete_dir : bytes -> unit Lwt.t
+
+  method virtual ensure_dir_exists : bytes -> sync : bool -> unit Lwt.t
+end
+
+class default_directory_access files_path =
+object(self)
+  inherit directory_access
+
+  val directory_cache : (string, directory_status) Hashtbl.t = Hashtbl.create 3
+
+  initializer
+    begin
+      let p = files_path in
+      if p.[0] <> '/'
+      then failwith (Printf.sprintf "'%s' should be an absolute path" p);
+      Hashtbl.add directory_cache "." Exists
+    end
+
+
+  method ensure_dir_exists dir ~sync =
+    match Hashtbl.find directory_cache dir with
+    | Exists -> Lwt.return ()
+    | Creating wait -> wait
+    | exception Not_found ->
+      let sleep, awake = Lwt.wait () in
+      (* the sleeper should be woken up under all
+         circumstances, hence the exception handling
+         below.
+         (otherwise this could e.g.
+          block the fragment cache...)
+       *)
+
+      Lwt.catch
+        (fun () ->
+           Hashtbl.add directory_cache dir (Creating sleep);
+
+           let parent_dir = Filename.dirname dir in
+           self # ensure_dir_exists parent_dir ~sync >>= fun () ->
+
+           Lwt_extra2.create_dir
+             ~sync
+             (Filename.concat files_path dir))
+        (function
+          | Unix.Unix_error (Unix.EEXIST, _, _) ->
+            Lwt.return ()
+          | exn ->
+            Hashtbl.remove directory_cache dir;
+            (* need to wake up the waiter here so it doesn't wait forever *)
+            Lwt.wakeup_exn awake exn;
+            Lwt.fail exn) >>= fun () ->
+
+      Hashtbl.replace directory_cache dir Exists;
+      Lwt.wakeup awake ();
+      Lwt.return_unit
+
+  method delete_dir = function
+    | "."
+    | "" -> Lwt.fail_with "just don't"
+    | dir ->
+      begin
+        let full_dir = Filename.concat files_path dir in
+        match Hashtbl.find directory_cache dir with
+        | Exists ->
+           Hashtbl.remove directory_cache dir;
+           Lwt_unix.rmdir full_dir
+        | Creating wait -> Lwt.fail_with (Printf.sprintf "creating %s" dir)
+        | exception Not_found ->
+                    Lwt_unix.rmdir full_dir
+      end
+
+end
+
+let _FILE_OPEN     = 40l
+let _FILE_CLOSE    = 41l
+let _READ_BATCH    = 42l
+let _READ_WAIT     = 43l
+let _READ_BATCH_CB = 44l
```

Comment:
Having these numbers in the same domain as a the asd commands is a bit of a hack ... I can live with it, but to prevent future accidents I think a note about this should be added near the Asd_protocol.ml command_map definition.